iscservice/demo/libdemo.c

```diff
@@ -0,0 +1,352 @@
+}
+
+int launch_stob_io(struct m0_isc_comp_private *pdata,
+		   struct m0_buf *in, int *rc)
+{
+	uint32_t           shift = 0;
+	struct m0_stob_io *stio = (struct m0_stob_io *)pdata->icp_data;
+	struct m0_fom     *fom = pdata->icp_fom;
+	struct isc_targs   ta = {};
+	struct m0_stob_id  stob_id;
+	struct m0_stob    *stob = NULL;
+
+	*rc = m0_xcode_obj_dec_from_buf(&M0_XCODE_OBJ(isc_targs_xc, &ta),
+					in->b_addr, in->b_nob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to xdecode args: rc=%d", *rc);
+		return M0_FSO_AGAIN;
+	}
+
+	if (ta.ist_ioiv.ci_nr == 0) {
+		M0_LOG(M0_ERROR, "no io segments given");
+		*rc = -EINVAL;
+		goto err;
+	}
+
+	m0_stob_io_init(stio);
+	stio->si_opcode = SIO_READ;
+
+	m0_fid_convert_cob2stob(&ta.ist_cob, &stob_id);
+	*rc = m0_storage_dev_stob_find(m0_cs_storage_devs_get(),
+				       &stob_id, &stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to find stob by cob="FID_F": rc=%d",
+		       FID_P(&ta.ist_cob), *rc);
+		goto err;
+	}
+
+	shift = m0_stob_block_shift(stob);
+
+	*rc = m0_indexvec_wire2mem(&ta.ist_ioiv, ta.ist_ioiv.ci_nr,
+				   shift, &stio->si_stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to make cob ivec: rc=%d", *rc);
+		goto err;
+	}
+
+	*rc = bufvec_alloc_init(&stio->si_user, &ta.ist_ioiv, shift);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to allocate bufvec: rc=%d", *rc);
+		goto err;
+	}
+
+	bufvec_pack(&stio->si_user, shift);
+
+	*rc = m0_stob_io_private_setup(stio, stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to setup adio for stob="FID_F": rc=%d",
+		       FID_P(&stob->so_id.si_fid), *rc);
+		goto err;
+	}
+
+	/* make sure the fom is waked up on I/O completion */
+	m0_mutex_lock(&stio->si_mutex);
+	m0_fom_wait_on(fom, &stio->si_wait, &fom->fo_cb);
+	m0_mutex_unlock(&stio->si_mutex);
+
+	*rc = m0_stob_io_prepare_and_launch(stio, stob, NULL, NULL);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to launch io for stob="FID_F": rc=%d",
+		       FID_P(&stob->so_id.si_fid), *rc);
+		m0_mutex_lock(&stio->si_mutex);
+		m0_fom_callback_cancel(&fom->fo_cb);
+		m0_mutex_unlock(&stio->si_mutex);
+	}
+ err:
+	if (*rc != 0) {
+		if (stob != NULL) {
+			bufvec_open(&stio->si_user, shift);
+			stio_fini(stio, stob);
+		}
+		/*
+		 * EAGAIN has a special meaning in the calling isc code,
+		 * so make sure we don't return it by accident.
+		 */
+		if (*rc == -EAGAIN)
+			*rc = -EBUSY;
+		return M0_FSO_AGAIN;
+	}
+
+	*rc = -EAGAIN; /* Wait for the I/O result. */
+	return M0_FSO_WAIT;
+}
+
+int compute_minmax(enum op op, struct m0_isc_comp_private *pdata,
+		   struct m0_buf *out, int *rc)
+{
+	int               i = 0;
+	int               n;
+	uint32_t          shift;
+	double            val;
+	FILE             *f;
+	char             *p;
+	struct mm_result  res = {};
+	struct m0_buf     buf = M0_BUF_INIT0;
+	struct m0_stob_io *stio = (struct m0_stob_io *)pdata->icp_data;
+
+	shift = m0_stob_block_shift(stio->si_obj);
+	bufvec_open(&stio->si_user, shift);
+	p = stio->si_user.ov_buf[0];
```

Comment:
Why don't we just treat this buffer as array of 32-bit float number?
So, e.g. 4096 bytes of binary buffer are just an array of 1024 32-bit float?

Reply:
or, 64-bit double?

Reply:
Because we want to show in this demo how to deal with the situations when the user data is not 4K aligned.